.. _MDS:

=============================
Multidimensional scaling plot
=============================

Summary
=======

Multidimensional scaling (MDS) is a way of transforming and visualising the 
datasets (similar to PCA analysis), to enable the identification of the strong 
patterns that account for the majority of the variation between the datasets. 

MDS analysis is closely related to PCA analysis, however the major difference is 
that MDS uses a distance matrix as input and minimises the dimensions of the 
data preserving the distance between data points, whilst PCA uses a correlation
(or variance-covariance) matrix as input and minimises the dimensons preserving 
the covariance of the data and finds the directions (components) that maximise 
the variance in the dataset. See `<http://www.analytictech.com/networks/mds.html>` 
for further details.

Results
=======

.. report:: RnaseqqcReport.sampleMDS
   :render: r-ggplot
   :statement: aes(x=MD1, y=MD2, colour=factor_value) +
	       geom_point() +
	       theme_bw() +
	       xlab('Dimension 1') + ylab ('Dimension 2') +
	       theme(
	       axis.text.x=element_text(size=20, angle=90, hjust=1, vjust=0.5),
	       axis.text.y=element_text(size=20))

   Plots of first and second dimensions from MDS analysis showing the
   latent variables that explain most of the variance in the dataset.

About this section
==================

Inputs
------

A table of expression values in TPM for each transcript and sample
generated by sailfish and stored as the 'transcript_quantification'
table in the rnaseqqc pipeline database

Outputs
-------

A scatter plot of the first two dimensions, dimension 1 on the x axis
and dimension 2 on the y axis. The way the datasets cluster represents
the factors driving the dissimlarities in the datasets.


How the results are generated
-----------------------------

1. Expression values are arranged in a dataframe organised such that there 
   is a row for each transcript and a columns for each sample.

2. Expression values are filtered to remove any genes that where TPM is 0 
   across all of the samples.

3. Euclidean distances are calulated amounf all pairs of points. 

4. MDS analysis is computed using sklearn.manifold.mds method
 
What you should expect from results
-----------------------------------

**Scatter Plot - Dimension 1 vs Dimension 2::

Ideally the first two dimensions (Dimension 1 and Dimension 2) should
seperate and cluster the datasets according to the experimental design
e.g. by tissue type or by treatment. If experimental design does not
seperate the datasets, this indicates there are other factors that are
having a greater influence on the expression values - these might be
things like batch-specific effects, sequencing lane, library prep or
other confounding factors and could affect the experimental results.
